cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkPackageUtils.java

```diff
@@ -462,29 +466,63 @@ private static SparkFramework prepareSparkFramework(SparkCompat sparkCompat,
+    File archive = new File(tempDir, archiveName);
+    try (
+      ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
+      zipOutput.setLevel(Deflater.NO_COMPRESSION);
+      for (File file : getLocalSparkLibrary(sparkCompat)) {
+        zipOutput.putNextEntry(new ZipEntry(file.getName()));
+        Files.copy(file.toPath(), zipOutput);
+        zipOutput.closeEntry();
+      }
+    }
 
-    if (!frameworkLocation.exists()) {
-      File archive = new File(tempDir, archiveName);
-      try {
-        try (ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
-          zipOutput.setLevel(Deflater.NO_COMPRESSION);
-          for (File file : getLocalSparkLibrary(sparkCompat)) {
-            zipOutput.putNextEntry(new ZipEntry(file.getName()));
-            Files.copy(file.toPath(), zipOutput);
-            zipOutput.closeEntry();
+    Location writeInProgressLoc = frameworkDir.append("write_in_progress");
+    //Create writing file
+    boolean isWriteComplete = false;
+    while (!isWriteComplete) {
+      if (writeInProgressLoc.createNew()) {
+        Location tempDirHdfs = locationFactory.create("/framework/spark_temp_" + UUID.randomUUID());
+        Location tempArchiveHdfs = tempDirHdfs.append(archiveName);
+        tempDirHdfs.mkdirs("755");
+        try {
+          //Copy file to temp loc ( Upload )
+          try (OutputStream os = tempArchiveHdfs.getOutputStream()) {
+            Files.copy(archive.toPath(), os);
           }
-        }
-
-        // Upload spark archive to the framework location
-        frameworkDir.mkdirs("755");
 
-        try (OutputStream os = frameworkLocation.getOutputStream("644")) {
-          Files.copy(archive.toPath(), os);
+          //Move to actual location
+          tempArchiveHdfs.renameTo(frameworkLocation);
+          writeInProgressLoc.delete();
+          isWriteComplete = true;
+        } finally {
+          if (tempArchiveHdfs.exists()) {
+            tempArchiveHdfs.delete();
+          }
+          archive.delete();
+        }
+      } else {
+        //In case of race condition 2nd process will wait for the 1st process to write
+        //Wait for the writeInProgressLoc to be deleted
+        try {
+          Tasks.waitFor(false, () -> writeInProgressLoc.exists(), 5, TimeUnit.MINUTES);
+          isWriteComplete = true;
+        } catch (TimeoutException e) {
+          //The spark jar copying was not completed by other process within the given timeout
+          //Clean the artifacts from other task and retry writing
+          writeInProgressLoc.delete();
+          if (frameworkLocation.exists()) {
+            frameworkLocation.delete();
+          }
+        } catch (ExecutionException | InterruptedException e) {
+          throw new IOException(e);
         }
-      } finally {
-        archive.delete();
       }
     }
+    
+    if(archive.exists()) {
```

Comment:
Use `Files.deleteIfExists`.

Reply:
Also, this should be in the finally block that covers starting at line 470.

Reply:
This is delete is for the case when 2 process write to local temp JAR. 
Then when trying to upload to HDFS , only 1 enters the `try` block to upload and Finally delete .

But the 2nd process will not enter the Upload try block, so this delete for this 2nd process.

Reply:
The problem is that the File was being created in the line 469-477 block. Any exception that happen after line 469 could left this temp file undeleted. That's why the common practice is to have a finally block to cover it to delete if it exists regardless of what happened.